doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,501 @@
+
+## Relevant data
+
+A (certainly non-exhaustive) selection of API design guides and
+implementations for the paging strategies outlined above can be found in
+the following links.
+
+* A broad discussion of paging strategies:
+  - [Patterns for API Design](https://microservice-api-patterns.org/patterns/quality/dataTransferParsimony/Pagination)
+* Offset based designs:
+  - [Atlassian](https://developer.atlassian.com/server/framework/atlassian-sdk/atlassian-rest-api-design-guidelines-version-1/#standard-query-parameters-in-uris)
+  - [Australian Government](https://apiguide.readthedocs.io/en/latest/build_and_publish/batching_results.html)
+  - [Microsoft](https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#982-client-driven-paging)
+* Page token based designs:
+  - [Google Cloud guidelines](https://cloud.google.com/apis/design/design_patterns#list_pagination)
+  - [Google Cloud Datastore](https://cloud.google.com/datastore/docs/concepts/queries#datastore-datastore-cursor-paging-python)
+  - [Twitter](https://developer.twitter.com/en/docs/twitter-api/pagination)
+  - [Slack](https://slack.engineering/evolving-api-pagination-at-slack/)
+* Cursor object designs:
+  - [Cisco](https://github.com/CiscoDevNet/api-design-guide#363-get)
+  - [Cloudfoundry](https://github.com/cloudfoundry/cc-api-v3-style-guide#pagination)
+  - [JDBC](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html)
+  - [MongoDB](https://www.mongodb.com/docs/manual/reference/method/js-cursor/)
+  - [Python PEP 239 DB-API](https://peps.python.org/pep-0249/#fetchmany)
+    used by many database libraries.
+  - [Ruby on Rails](https://api.rubyonrails.org/v7.0.4.2/classes/ActiveRecord/Batches.html#method-i-find_in_batches)
+  - [W3C IndexedDB Cursor API spec](https://w3c.github.io/IndexedDB/#idbcursor)
+
+### Rejected options
+
+In order to avoid considering a long list of options that are either
+clearly inferior, or broadly similar to other options, some valid
+solutions were rejected early.
+
+#### Whole-batch offset
+
+The classic offset-based solution was rejected due to the critical
+drawback of unstable paging, which a page token based solution overcomes
+without introducing much more complexity.
+
+```python
+SuccessCallback = Callable[[int, List[EntityReferences]], None]
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship,
+        successCb: SuccessCallback,
+        errorCb,
+        limit: int = -1, offset: int = 0): ...
+```
+
+##### Pros
+
+- Conceptually simple for the host.
+- Jumping arbitrarily to nth page is trivial.
+- Retains batch-awareness for each page.
+- Transactions, indexing and caching technologies exist that can
+  alleviate stability and performance problems, to some extent.
+
+##### Cons
+
+- Anomalous page results if insertions/deletions can happen before next
+  page is retrieved, or if ordering is not stable.
+- Server cannot negotiate page size - there is no mechanism to signal
```

Comment:
Added page size negotiation Con after offline conversation (also corresponding Pro for page token) in 042c17a5